File: solutions/csharp/run-length-encoding/1/RunLengthEncoding.cs
using System.Text.RegularExpressions;

public static class RunLengthEncoding
{
    private const string EncodedRunPattern = @"(\d*)(\D)";

    public static string Encode(string input) =>
        string.Concat(Runs(input).Select(EncodeRun));

    public static string Decode(string input) =>
        Regex.Replace(input, EncodedRunPattern, DecodeRun);

    private static string EncodeRun((char Value, int Count) run) =>
        run.Count == 1 ? run.Value.ToString() : $"{run.Count}{run.Value}";

    private static string DecodeRun(Match match)
    {
        var countText = match.Groups[1].Value;
        var value = match.Groups[2].Value[0];
        var count = string.IsNullOrEmpty(countText) ? 1 : int.Parse(countText);

        return new string(value, count);
    }

    private static IEnumerable<(char Value, int Count)> Runs(string input)
    {
        if (input.Length == 0)
        {
            yield break;
        }

        var current = input[0];
        var count = 1;

        foreach (var value in input[1..])
        {
            if (value == current)
            {
                count++;
            }
            else
            {
                yield return (current, count);
                current = value;
                count = 1;
            }
        }

        yield return (current, count);
    }
}

File: solutions/scala/run-length-encoding/1/src/main/scala/RunLengthEncoding.scala
import scala.util.matching.Regex

object RunLengthEncoding {
  private val Run = """(.)\1*""".r
  private val EncodedRun = """(\d+)?(\D)""".r

  def encode(input: String): String =
    Run
      .findAllIn(input)
      .map(encodeRun)
      .mkString

  def decode(input: String): String =
    EncodedRun
      .findAllMatchIn(input)
      .map(expand)
      .mkString

  private def encodeRun(run: String): String =
    if run.length == 1 then run
    else s"${run.length}${run.head}"

  private def expand(encodedRun: Regex.Match): String =
    val count = Option(encodedRun.group(1)).fold(1)(_.toInt)
    encodedRun.group(2) * count
}
